### ISSUE TITLE
redirect loop when accessing account management console

### ISSUE THREAD

**Issue by Degoya-Studio** [keycloak/keycloak#34292]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

account/ui

### Describe the bug

I'm trying to run KeyCloak on Digital Ocean App Platform and everything works so far, except for the account console. When I try to open it for a none master realm, it ends up in an infinite redirect loop. I've tried a lot of things out but to not avail.

**Digital Ocean App Spec**

```
alerts:
  - rule: DEPLOYMENT_FAILED
  - rule: DOMAIN_FAILED
domains:
  - domain: auth.domain.dzg.de
    type: PRIMARY
    zone: dzg.de
features:
  - buildpack-stack=ubuntu-22
ingress:
  rules:
    - component:
        name: dzg-auth-keycloak
      match:
        path:
          prefix: /
name: dzg-domain-keycloak-server
region: fra
services:
  - dockerfile_path: Dockerfile
    envs:
      - key: KC_DB_URL
        scope: RUN_AND_BUILD_TIME
        value: jdbc:postgresql://dzg-domain-db-cluster-postgresql-do-user-17085909-0.i.db.ondigitalocean.com:25060/keycloak
      - key: KC_DB_USERNAME
        scope: RUN_AND_BUILD_TIME
        value: doadmin
      - key: KC_DB_PASSWORD
        scope: RUN_AND_BUILD_TIME
        type: SECRET
        value: EV[1:KgX0LcruXm/Gu7DVca62/55uXMklwToK:c+XSu0GiEhA33zWHZ8gKYfSqtPHAiPfILdI9fisAbfX6zdQ9nQsFOw==]
      - key: KEYCLOAK_ADMIN
        scope: RUN_AND_BUILD_TIME
        value: admin
      - key: KEYCLOAK_ADMIN_PASSWORD
        scope: RUN_AND_BUILD_TIME
        value: password
      - key: KC_HOSTNAME
        scope: RUN_AND_BUILD_TIME
        value: auth.domain.dzg.de
      - key: KC_SPI_EVENTS_LISTENER_EVENT_LISTENER_API_URL
        scope: RUN_AND_BUILD_TIME
        value: https://api.domain.dzg.de/api/v1
      - key: KC_SPI_EVENTS_LISTENER_EVENT_LISTENER_API_KEY
        scope: RUN_AND_BUILD_TIME
        type: SECRET
        value: EV[1:v7DRsuEk9O2sQ3L8Yrcd6ymDR9J9uQ0D:RQ7lgyPa3+FjtzECLIcXZWF51/trTOxxmT6AVCTFnBfVDO0=]
      - key: KC_PROXY
        scope: RUN_AND_BUILD_TIME
        value: edge
      - key: KC_PROXY_ADDRESS_FORWARDING
        scope: RUN_AND_BUILD_TIME
        value: "true"
      - key: KC_PROXY_HEADERS
        scope: RUN_AND_BUILD_TIME
        value: xforwarded
      - key: KC_HOSTNAME_STRICT
        scope: RUN_AND_BUILD_TIME
        value: "false"
      - key: KC_HTTP_ENABLED
        scope: RUN_AND_BUILD_TIME
        value: "false"
      - key: KC_HOSTNAME_STRICT_HTTPS
        scope: RUN_AND_BUILD_TIME
        value: "false"
      - key: KC_SPI_COOKIE_DEFAULT_SECURE
        scope: RUN_AND_BUILD_TIME
        value: "false"
      - key: KC_SPI_COOKIE_DEFAULT_SAME_SITE
        scope: RUN_AND_BUILD_TIME
        value: "false"
      - key: KC_HTTP_RELATIVE_PATH
        scope: RUN_AND_BUILD_TIME
        value: /
      - key: KC_HOSTNAME_DEBUG
        scope: RUN_AND_BUILD_TIME
        value: "true"
      - key: KC_HOSTNAME_BACKCHANNEL_DYNAMIC
        scope: RUN_AND_BUILD_TIME
        value: "false"
    github:
      branch: main
      deploy_on_push: true
      repo: Degoya-Studio/dzg-auth-keycloak
    http_port: 8080
    instance_count: 1
    instance_size_slug: apps-s-1vcpu-1gb
    name: dzg-auth-keycloak
    source_dir: /
```

**Keycloak Export files**

[account.json](https://github.com/user-attachments/files/17505355/account.json)
[realm-export.json](https://github.com/user-attachments/files/17505447/realm-export.json)

**Screenshots**

<img width="1785" alt="Bildschirmfoto 2024-10-24 um 11 54 27" src="https://github.com/user-attachments/assets/ddbfe55d-78e8-436d-b8fc-0f96eaee7970">
<img width="1761" alt="Bildschirmfoto 2024-10-24 um 11 54 47" src="https://github.com/user-attachments/assets/34c85e69-39f3-4ec4-821d-09680d1e6ad2">
<img width="645" alt="Bildschirmfoto 2024-10-24 um 12 02 49" src="https://github.com/user-attachments/assets/9bd506d4-e693-43f9-9b12-93ac9abb6d15">

### Version

26.0.0

### Regression

- [ ] The issue is a regression

### Expected behavior

Access the Account Management Console.

### Actual behavior

redirect loop when accessing the Account Management Console for a non master realm

### How to Reproduce?

Run Keycloak on Digital Ocean App Platform and open https://<url>/realms/<none-master>/account

### Anything else?

_dzg is not the actual name of the domain. It's just a realm name_

https://github.com/keycloak/keycloak/issues/33109

**Comment by sterzuccio** [comment]
I'm encountering the same problem with the AWS service. Has a solution been found yet?

**Comment by Degoya-Studio** [comment]
I've read through these pages but wasn't able to solve it - [server hostname](https://www.keycloak.org/server/hostname),  [server reverseproxy](
https://www.keycloak.org/server/reverseproxy)

**Comment by Degoya-Studio** [comment]
sure **https://auth.domain.dzg.de/realms/dzg/account** about the proxy, Digital Ocean gives you no option to set it, so I do not know how they have configured it under the hood. I have applied all of your recommendations  but unfortunately it did not solve my problem.

**Docker File**

```

FROM maven:3.9-eclipse-temurin-17 AS builder-maven

COPY pom.xml /build/
COPY src /build/src/

WORKDIR /build
RUN mvn clean package

FROM quay.io/keycloak/keycloak:26.0 as builder-keycloak

ENV KC_HEALTH_ENABLED=true
ENV KC_METRICS_ENABLED=true

ENV KC_DB=postgres
ENV KC_FEATURES=token-exchange

ENV KC_CACHE=local
ENV KC_CACHE_STACK=local
ENV KC_CACHE_CONFIG_FILE=cache-local.xml

COPY --from=builder-maven /build/target/event-listener.jar /opt/keycloak/providers/
COPY themes/. /opt/keycloak/themes/.

RUN /opt/keycloak/bin/kc.sh build --http-relative-path=/

FROM quay.io/keycloak/keycloak:26.0
COPY --from=builder-keycloak /opt/keycloak/ /opt/keycloak/

ENV KC_DB_URL=jdbc:postgresql://${POSTGRES_DB}?ssl=true&sslmode=verify-full
ENV KC_DB_USERNAME=${POSTGRES_USER}
ENV KC_DB_PASSWORD=${POSTGRES_PASSWORD}

ENV KC_BOOTSTRAP_ADMIN_USERNAME=admin
ENV KC_BOOTSTRAP_ADMIN_PASSWORD=password
ENV KC_SPI_EVENTS_LISTENER_EVENT_LISTENER_API_URL=${INGE_API_URL}
ENV KC_SPI_EVENTS_LISTENER_EVENT_LISTENER_API_KEY=${INGE_API_KEY}

COPY --from=builder-keycloak /opt/keycloak/providers/ /opt/keycloak/providers/
COPY --from=builder-keycloak /opt/keycloak/themes/ /opt/keycloak/themes/

EXPOSE 8443

ENV JAVA_OPTS_APPEND="-Djava.net.preferIPv4Stack=true -Dinfinispan.cluster.enabled=false"


ENTRYPOINT ["/opt/keycloak/bin/kc.sh", "start", "--optimized", \
    "--import-realm"]
```

**Environment Variables**

```
KC_DB_URL=jdbc:postgresql://dzg-domain-db-cluster-postgresql-do-user-19085703-0.i.db.ondigitalocean.com:25060/keycloak
KC_DB_USERNAME=doadmin
KC_DB_PASSWORD=********************
BOOTSTRAP_ADMIN_USERNAME=admin
BOOTSTRAP_ADMIN_PASSWORD=password
KC_HOSTNAME=auth.domain.dzg.de
KC_SPI_EVENTS_LISTENER_EVENT_LISTENER_API_URL=https://api.domain.dzg.de/api/v1
KC_SPI_EVENTS_LISTENER_EVENT_LISTENER_API_KEY=********************
KC_HTTP_ENABLED=true
KC_SPI_COOKIE_DEFAULT_SECURE=false
KC_SPI_COOKIE_DEFAULT_SAME_SITE=false
KC_HTTP_RELATIVE_PATH=/
KC_HOSTNAME_DEBUG=true
KC_HOSTNAME_BACKCHANNEL_DYNAMIC=false
KC_PROXY_HEADERS=xforwarded
PROXY_PROTOCOL_ENABLED=true
```

**Comment by shawkins** [comment]
I tried to reproduce your setup, but haven't yet reproduced a redirect loop.

The most common reason keycloak would respond with a redirect is if there is an error handling that auth url - turn on event logging or look at the response location header to see if there is anything like that happening. 

A few more configuration notes:

KC_HTTP_ENABLED=true doesn't need to be enabled since the container only has EXPOSE 8443, which is the https port. 

BOOTSTRAP_ADMIN_USERNAME/PASSWORD, and PROXY_PROTOCOL_ENABLED need to have KC_ prefixes. Also proxy protocol is only useful if you are doing passthrough https.

**Comment by Degoya-Studio** [comment]
@shawkins I still don't know what caused it, but I ended up creating a new realm and it works there. Thank you again for your help.

I will still provide a log in case you are able to infer to find something
[keycloak.log](https://github.com/user-attachments/files/17527651/keycloak.log)

**Comment by shawkins** [comment]
That's good news, are you ok with closing this issue?

**Comment by adedjoumaaboudalla** [comment]
Ran into this exact same issue (infinite redirect loop on account console for a non-master realm, `refresh_token` ending up `undefined` on repeated `/token` calls). Took a while to track down since the server was returning perfectly valid tokens (confirmed via manual `curl` replay of `grant_type=refresh_token`), so it looked like a client-side bug at first.

For us it turned out to be **two separate realm-level misconfigurations**, neither of which threw a clear error:

1. **`ssoSessionMaxLifespan` was `0`** (shown as a blank field in the admin UI under Realm settings → Sessions → "SSO Session Max", but exports as `0` in the partial export JSON). We compared a partial realm export against a freshly created realm (which worked fine) and this was one of the two diffs. Setting it to a real value fixed part of the issue.

2. **Missing `clientScopeMappings`** for the `account` client scope → `account-console` client. A working realm has this by default:
   ```json
   "clientScopeMappings": {
     "account": [
       {
         "client": "account-console",
         "roles": ["manage-account", "view-groups"]
       }
     ]
   }
   ```
   Ours was entirely absent (not just empty — the key didn't exist at all in the export). Without `manage-account`, calls from the account console's React app to the Account REST API likely returned 403, which seems to get misinterpreted by the client as an invalid session, triggering a fresh re-auth attempt — which then loops.

   To check/fix this in the admin UI: **Clients → account-console → "Client scopes" tab → click on `account-console-dedicated` → "Scope" tab**. If `manage-account` / `view-groups` aren't listed there, add them via "Assign role" → filter by clients → `account`.

We never figured out *how* our realm ended up in that state (possibly a partial import at some point, or a manual edit that silently dropped the default scope mapping). Recreating the `account-console` client alone (delete + reimport) did **not** fix it — the missing piece was the client scope mapping.

If anyone else hits this: export your realm (`Partial export`, all options checked), diff it against a fresh realm created with default settings, and check both of these fields specifically.
